# Exercício de Programação 5: Regressão

<font color="red">**Prazo de submissão: 23:55 do dia 16/03/2021** </font>

2020.2 Álgebra Linear Computacional - DCC - UFMG

Erickson - Fabricio

Instruções:
* Antes de submeter suas soluções, certifique-se de que tudo roda como esperado. Primeiro, **reinicie o kernel** no menu, selecione Kernel$\rightarrow$Restart e então execute **todas as células** (no menu, Cell$\rightarrow$Run All)
* Apenas o arquivo .ipynb deve ser submetido. Ele não deve ser compactado.
* Não deixe de preencher seu nome e número de matrícula na célula a seguir

**Nome do aluno:** Arthur Pontes Nader

**Matricula:** 2019022294

## Carregando os dados

Iremos carregar os dados usando a biblioteca ```pandas```. Não se preocupe se você não conhece a biblioteca, pois o nosso objetivo é apenas extrair a matriz de dados $X$. Segue uma descrição do dataset, retirada [daqui](http://statweb.stanford.edu/~owen/courses/202/Cereals.txt).

* Datafile Name: Cereals
* Datafile Subjects: Food , Health
* Story Names: Healthy Breakfast
* Reference: Data available at many grocery stores
* Authorization: free use
* Description: Data on several variable of different brands of cereal.

A value of -1 for nutrients indicates a missing observation.
Number of cases: 77
Variable Names:

  1. Name: Name of cereal
  2. mfr: Manufacturer of cereal where A = American Home Food Products; G =
     General Mills; K = Kelloggs; N = Nabisco; P = Post; Q = Quaker Oats; R
     = Ralston Purina
  3. type: cold or hot
  4. calories: calories per serving
  5. protein: grams of protein
  6. fat: grams of fat
  7. sodium: milligrams of sodium
  8. fiber: grams of dietary fiber
  9. carbo: grams of complex carbohydrates
  10. sugars: grams of sugars
  11. potass: milligrams of potassium
  12. vitamins: vitamins and minerals - 0, 25, or 100, indicating the typical percentage of FDA recommended
  13. shelf: display shelf (1, 2, or 3, counting from the floor)
  14. weight: weight in ounces of one serving
  15. cups: number of cups in one serving
  16. rating: a rating of the cereals

In [1]:
#Execute esta célula para instalar o pandas caso já não tenha instalado
# import sys
# !{sys.executable} -m pip install --user pandas

In [2]:
import pandas as pd
df = pd.read_table('cereal.txt',sep='\s+',index_col='name')
df

mfr type  calories  protein  fat  sodium  fiber  \
name                                                                        
100%_Bran                   N    C        70        4    1     130   10.0   
100%_Natural_Bran           Q    C       120        3    5      15    2.0   
All-Bran                    K    C        70        4    1     260    9.0   
All-Bran_with_Extra_Fiber   K    C        50        4    0     140   14.0   
Almond_Delight              R    C       110        2    2     200    1.0   
...                        ..  ...       ...      ...  ...     ...    ...   
Triples                     G    C       110        2    1     250    0.0   
Trix                        G    C       110        1    1     140    0.0   
Wheat_Chex                  R    C       100        3    1     230    3.0   
Wheaties                    G    C       100        3    1     200    3.0   
Wheaties_Honey_Gold         G    C       110        2    1     200    1.0   

                           carbo  sugars  potass  vitamins  shelf  weight  \
name                                                                        
100%_Bran                    5.0       6     280        25      3     1.0   
100%_Natural_Bran            8.0       8     135         0      3     1.0   
All-Bran                     7.0       5     320        25      3     1.0   
All-Bran_with_Extra_Fiber    8.0       0     330        25      3     1.0   
Almond_Delight              14.0       8      -1        25      3     1.0   
...                          ...     ...     ...       ...    ...     ...   
Triples                     21.0       3      60        25      3     1.0   
Trix                        13.0      12      25        25      2     1.0   
Wheat_Chex                  17.0       3     115        25      1     1.0   
Wheaties                    17.0       3     110        25      1     1.0   
Wheaties_Honey_Gold         16.0       8      60        25      1     1.0   

                           cups     rating  
name                                        
100%_Bran                  0.33  68.402973  
100%_Natural_Bran          1.00  33.983679  
All-Bran                   0.33  59.425505  
All-Bran_with_Extra_Fiber  0.50  93.704912  
Almond_Delight             0.75  34.384843  
...                         ...        ...  
Triples                    0.75  39.106174  
Trix                       1.00  27.753301  
Wheat_Chex                 0.67  49.787445  
Wheaties                   1.00  51.592193  
Wheaties_Honey_Gold        0.75  36.187559  

[77 rows x 15 columns]

A seguir iremos remover as linhas correspondentes aos cereais que possuem dados faltantes, representados pelo valor -1.
Também iremos remover as colunas com dados categóricos 'mfr' e 'type', e os dados numéricos, 'shelf', 'weight' e 'cups'.

In [3]:
import numpy as np
new_df = df.replace(-1,np.nan)
new_df = new_df.dropna()
new_df = new_df.drop(['mfr','type','shelf','weight','cups'],axis=1)
new_df

calories  protein  fat  sodium  fiber  carbo  \
name                                                                      
100%_Bran                        70        4    1     130   10.0    5.0   
100%_Natural_Bran               120        3    5      15    2.0    8.0   
All-Bran                         70        4    1     260    9.0    7.0   
All-Bran_with_Extra_Fiber        50        4    0     140   14.0    8.0   
Apple_Cinnamon_Cheerios         110        2    2     180    1.5   10.5   
...                             ...      ...  ...     ...    ...    ...   
Triples                         110        2    1     250    0.0   21.0   
Trix                            110        1    1     140    0.0   13.0   
Wheat_Chex                      100        3    1     230    3.0   17.0   
Wheaties                        100        3    1     200    3.0   17.0   
Wheaties_Honey_Gold             110        2    1     200    1.0   16.0   

                           sugars  potass  vitamins     rating  
name                                                            
100%_Bran                     6.0   280.0        25  68.402973  
100%_Natural_Bran             8.0   135.0         0  33.983679  
All-Bran                      5.0   320.0        25  59.425505  
All-Bran_with_Extra_Fiber     0.0   330.0        25  93.704912  
Apple_Cinnamon_Cheerios      10.0    70.0        25  29.509541  
...                           ...     ...       ...        ...  
Triples                       3.0    60.0        25  39.106174  
Trix                         12.0    25.0        25  27.753301  
Wheat_Chex                    3.0   115.0        25  49.787445  
Wheaties                      3.0   110.0        25  51.592193  
Wheaties_Honey_Gold           8.0    60.0        25  36.187559  

[74 rows x 10 columns]

Finalmente, iremos converter os dados nutricionais numéricos de ```new_df``` para uma matriz ```dados``` e as avaliações (ratings) para um vetor $y$. Os nomes dos cereais serão salvos em uma lista ```cereral_names``` e os nomes das colunas em uma lista ```col_names```.

In [4]:
cereral_names = list(new_df.index)
print('Cereais:',cereral_names)
col_names = list(new_df.columns)
print('Colunas:',col_names)

dados = new_df.drop('rating', axis=1).values
print('As dimensões de dados são:',dados.shape)
y = new_df['rating'].values
print('As dimensões de y são:',y.shape)

Cereais: ['100%_Bran', '100%_Natural_Bran', 'All-Bran', 'All-Bran_with_Extra_Fiber', 'Apple_Cinnamon_Cheerios', 'Apple_Jacks', 'Basic_4', 'Bran_Chex', 'Bran_Flakes', "Cap'n'Crunch", 'Cheerios', 'Cinnamon_Toast_Crunch', 'Clusters', 'Cocoa_Puffs', 'Corn_Chex', 'Corn_Flakes', 'Corn_Pops', 'Count_Chocula', "Cracklin'_Oat_Bran", 'Crispix', 'Crispy_Wheat_&_Raisins', 'Double_Chex', 'Froot_Loops', 'Frosted_Flakes', 'Frosted_Mini-Wheats', 'Fruit_&_Fibre_Dates,_Walnuts,_and_Oats', 'Fruitful_Bran', 'Fruity_Pebbles', 'Golden_Crisp', 'Golden_Grahams', 'Grape_Nuts_Flakes', 'Grape-Nuts', 'Great_Grains_Pecan', 'Honey_Graham_Ohs', 'Honey_Nut_Cheerios', 'Honey-comb', 'Just_Right_Crunchy__Nuggets', 'Just_Right_Fruit_&_Nut', 'Kix', 'Life', 'Lucky_Charms', 'Maypo', 'Muesli_Raisins,_Dates,_&_Almonds', 'Muesli_Raisins,_Peaches,_&_Pecans', 'Mueslix_Crispy_Blend', 'Multi-Grain_Cheerios', 'Nut&Honey_Crunch', 'Nutri-Grain_Almond-Raisin', 'Nutri-grain_Wheat', 'Oatmeal_Raisin_Crisp', 'Post_Nat._Raisin_Bran', 'Prod

## Estimando os parâmetros da regressão linear simples

Qual será a relação entre a avaliação $y$ e o número de calorias $x$ de um cereal? Para responder esta pergunta, considere uma regressão linear simples
$$
y = \beta_0 + \beta_1 x.
$$
Para encontrar os coeficientes $\beta_0$ e $\beta_1$ utilizando o método dos mínimos quadrados, basta resolver o sistema
$$
\begin{bmatrix}
n & \sum_i x^{(i)} \\
\sum_i x^{(i)} & \sum_i (x^{(i)})^2
\end{bmatrix}
\begin{bmatrix}
\beta_0 \\ \beta_1
\end{bmatrix}
=
\begin{bmatrix}
\sum_i y^{(i)} \\ \sum_i x^{(i)} y^{(i)}
\end{bmatrix}
$$

Portanto, para encontrar $\beta_0$ e $\beta_1$, você precisa
1. Calcular a matriz
$$
A = \begin{bmatrix}
n & \sum_i x^{(i)} \\
\sum_i x^{(i)} & \sum_i (x^{(i)})^2
\end{bmatrix}
$$
e o vetor
$$
c = \begin{bmatrix}
\sum_i y^{(i)} \\ \sum_i x^{(i)} y^{(i)}
\end{bmatrix}
$$
2. Resolver $A \beta = c$, onde $\beta$ é o vetor de coeficientes.

**Exercício 1 - Regressão simples:** Encontre os coeficientes $\beta_0$ e $\beta_1$ quando a variável independente é ```calories```. Dica: A variavel X abaixo já armazena os valores deste atributo.

In [5]:
X = new_df['calories'].values

# insira seu código aqui

A = np.array([[X.shape[0], X.sum()],[X.sum(), X@X.T]])

c = np.array([[y.sum()],[y@X]])

beta = np.linalg.inv(A) @ c
print(beta)

[[94.88442777]
 [-0.49064841]]


**Exercício 2 - Regressão múltipla:** Considerando a nova tabela de dados X abaixo com os atributos 'calories', 'protein', 'fat', 'sugars' e 'vitamins' selecionados, estime os parâmetros da regressão múltipla para obter a variavel resposta ```rating``` 

In [6]:
X = new_df.loc[:,['calories', 'protein', 'fat', 'sugars', 'vitamins']].values

# insira seu código aqui

ones = np.ones(X.shape[0]).reshape(X.shape[0],1)

X = np.concatenate((ones,X), axis=1)

rating = (np.linalg.inv(X.T @ X) @ X.T @ y).reshape(6,1)
print(rating)

[[69.6495031 ]
 [-0.2618972 ]
 [ 5.55855063]
 [-2.8585006 ]
 [-1.14383072]
 [-0.07675886]]


**Exercício 3:** Nossos modelos de regressão linear são bons preditores da nota de avaliação do cereal? Qual o melhor modelo? Calcule os coeficientes de determinação e faça uma análise dos valores obtidos para responder a estas perguntas.

In [7]:
# insira seu código aqui

delta_y_simples = (y - np.array(new_df['calories'].values*beta[1] + beta[0]))**2

coef_reg_simples = 1 - np.sum(delta_y_simples)/np.sum((y - np.mean(y))**2)

delta_y_multipla = (y.reshape(74,1) - X@rating)**2

coef_reg_multipla = 1 - np.sum(delta_y_multipla)/np.sum((y - np.mean(y))**2)

print("Coeficiente de determinacao da regressao linear simples: " , coef_reg_simples, "\n")
print("Coeficiente de determinacao da regressao linear multipla: " , coef_reg_multipla)


Coeficiente de determinacao da regressao linear simples:  0.4813371585498133 

Coeficiente de determinacao da regressao linear multipla:  0.8337963187084577


Comparando-se os coeficientes de determinação, é possível perceber que a regressão linear simples apresentou um coeficiente bastante distante de 1 (0.481), o que indica que o modelo não é muito bom preditor da nota de avaliação do cereal. Já a regressão linear múltipla apresentou um coeficiente bem mais perto de 1 (0.834). Assim, pode-se concluir que o o modelo obtido pela regressão linear múltipla foi melhor pois ele se ajustou bem aos dados coletados, o que implica em uma predição mais significativa da avalição de um novo cereal de acordo com as suas características nutricionais.